src/lib/mixer_module/mixer_module.cpp

```diff
@@ -397,12 +468,56 @@ bool MixingOutput::update()
 	}
 
 	/* do mixing */
-	float outputs[MAX_ACTUATORS] {};
-	const unsigned mixed_num_outputs = _mixers->mix(outputs, _max_num_outputs);
+	float mixer_outputs[MAX_ACTUATORS] {};
+
+	unsigned mixed_num_outputs = 0;
+
+	if (_mixers) {
+		mixed_num_outputs = _mixers->mix(mixer_outputs, _max_num_outputs);
+	}
+
+	// Mask of all channels to be output; initialze to all mixed channels
+	unsigned mixed_outputs_mask = (1 << mixed_num_outputs) - 1;
+
+	/* get output controls for required topics */
+
+	for (unsigned grp = n_act; grp < n_act + n_out; grp++) {
+		if (_groups_subscribed & (1 << grp)) {
+			output_control_s controls;
+
+			if (_control_subs[grp].copy(&controls)) {
+				n_updates++;
+
+				for (uint8_t i = 0; i < controls.n_outputs; i++) {
+					auto func = controls.function[i];
+
+					if (func == 0) {
+						continue;
+					}
+
+					// Map the current control function to the correct output(s)
+					for (uint8_t j = 0; j < _max_num_outputs; j++) {
+						if (func == _assigned_function[j]) {
```

Comment:
Related to my comment above. If two control_output topics set the same function, the physical output will be written a value that depends on publish time and subscription order (it may even flicker between the two setpoints? 😳). 
IMO the 'function' field in output_control is redundant and I suggest keeping params only.

Reply:
You could say the same thing about the `actuator_controls` topics right now.  Different modules can publish to different instances of `actuator_controls`, you just have to trust the developer to use the correct instance.

Haven't figured out all the details of this yet, but I want it to be as flexible as possible.

One thing that's still missing is metadata about what kind of actuator is attached - i.e. ESC or servo.  I'm not sure if we could make use of the function field to determine that, but we need a way to determine whether a specific output is an ESC, servo, or some other device, so that we can do things like calibration, and intelligent feedback and failure monitoring.

Reply:
Well not quite, if you publish on the same control topic concurently in two modules, you are explicitly doing something wrong.
Here you can publish in two different canal and yet get a conflict, this is much more insidious and in my opinion uneeded.